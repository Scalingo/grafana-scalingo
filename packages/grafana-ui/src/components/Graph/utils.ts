import { GraphSeriesValue, Field } from '@grafana/data';

/**
 * Returns index of the closest datapoint BEFORE hover position
 *
 * @param posX
 * @param series
 */
export const findHoverIndexFromData = (xAxisDimension: Field, xPos: number) => {
  let lower = 0;
  let upper = xAxisDimension.values.length - 1;
  let middle;

  while (true) {
    if (lower > upper) {
      return Math.max(upper, 0);
    }
    middle = Math.floor((lower + upper) / 2);
    const xPosition = xAxisDimension.values.get(middle);

    if (xPosition === xPos) {
      return middle;
    } else if (xPosition && xPosition < xPos) {
      lower = middle + 1;
    } else {
      upper = middle - 1;
    }
  }
};

interface MultiSeriesHoverInfo {
  value: string;
  time: string;
  datapointIndex: number;
  seriesIndex: number;
  label?: string;
  color?: string;
}

/**
 * Returns information about closest datapoints when hovering over a Graph
 *
 * @param seriesList list of series visible on the Graph
 * @param pos mouse cursor position, based on jQuery.flot position
 */
export const getMultiSeriesGraphHoverInfo = (
  // x and y axis dimensions order is aligned
  yAxisDimensions: Field[],
  xAxisDimensions: Field[],
  /** Well, time basically */
  xAxisPosition: number
): {
  results: MultiSeriesHoverInfo[];
  time?: GraphSeriesValue;
} => {
  let value, i, series, hoverIndex, hoverDistance, pointTime;

  const results: MultiSeriesHoverInfo[] = [];

  let minDistance, minTime;

  for (i = 0; i < yAxisDimensions.length; i++) {
    series = yAxisDimensions[i];
    const time = xAxisDimensions[i];
    hoverIndex = findHoverIndexFromData(time, xAxisPosition);
    hoverDistance = xAxisPosition - time.values.get(hoverIndex);
    pointTime = time.values.get(hoverIndex);
    // Take the closest point before the cursor, or if it does not exist, the closest after
    if (
      minDistance === undefined ||
      (hoverDistance >= 0 && (hoverDistance < minDistance || minDistance < 0)) ||
      (hoverDistance < 0 && hoverDistance > minDistance)
    ) {
      minDistance = hoverDistance;
      minTime = time.display ? time.display(pointTime).text : pointTime;
    }

    value = series.values.get(hoverIndex);

    results.push({
      value: series.display ? series.display(value).text : value,
      datapointIndex: hoverIndex,
      seriesIndex: i,
      color: series.config.color,
      label: series.name,
      time: time.display ? time.display(pointTime).text : pointTime,
    });
  }

  return {
    results,
    time: minTime,
  };
};
