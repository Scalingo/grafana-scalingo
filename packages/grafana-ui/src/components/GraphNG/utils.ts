import { ArrayVector, DataFrame, Field, FieldConfig, FieldType, outerJoinDataFrames, TimeRange } from '@grafana/data';
import {
  AxisPlacement,
  GraphDrawStyle,
  GraphFieldConfig,
  ScaleDistribution,
  ScaleDistributionConfig,
} from '@grafana/schema';

import { FIXED_UNIT } from './GraphNG';
import { applyNullInsertThreshold } from './nullInsertThreshold';
import { nullToUndefThreshold } from './nullToUndefThreshold';
import { XYFieldMatchers } from './types';

function isVisibleBarField(f: Field) {
  return (
    f.type === FieldType.number && f.config.custom?.drawStyle === GraphDrawStyle.Bars && !f.config.custom?.hideFrom?.viz
  );
}

// will mutate the DataFrame's fields' values
function applySpanNullsThresholds(frame: DataFrame) {
  let refField = frame.fields.find((field) => field.type === FieldType.time); // this doesnt need to be time, just any numeric/asc join field
  let refValues = refField?.values.toArray() as any[];

  for (let i = 0; i < frame.fields.length; i++) {
    let field = frame.fields[i];

    if (field === refField || isVisibleBarField(field)) {
      continue;
    }

    let spanNulls = field.config.custom?.spanNulls;

    if (typeof spanNulls === 'number') {
      if (spanNulls !== -1) {
        field.values = new ArrayVector(nullToUndefThreshold(refValues, field.values.toArray(), spanNulls));
      }
    }
  }

  return frame;
}

export function preparePlotFrame(frames: DataFrame[], dimFields: XYFieldMatchers, timeRange?: TimeRange | null) {
  // apply null insertions at interval
  frames = frames.map((frame) => applyNullInsertThreshold(frame, null, timeRange?.to.valueOf()));

  let numBarSeries = 0;

  frames.forEach((frame) => {
    frame.fields.forEach((f) => {
      if (isVisibleBarField(f)) {
        // prevent minesweeper-expansion of nulls (gaps) when joining bars
        // since bar width is determined from the minimum distance between non-undefined values
        // (this strategy will still retain any original pre-join nulls, though)
        f.config.custom = {
          ...f.config.custom,
          spanNulls: -1,
        };

        numBarSeries++;
      }
    });
  });

  // to make bar widths of all series uniform (equal to narrowest bar series), find smallest distance between x points
  let minXDelta = Infinity;

  if (numBarSeries > 1) {
    frames.forEach((frame) => {
      if (!frame.fields.some(isVisibleBarField)) {
        return;
      }

      const xVals = frame.fields[0].values.toArray();

      for (let i = 0; i < xVals.length; i++) {
        if (i > 0) {
          minXDelta = Math.min(minXDelta, xVals[i] - xVals[i - 1]);
        }
      }
    });
  }

  let alignedFrame = outerJoinDataFrames({
    frames,
    joinBy: dimFields.x,
    keep: dimFields.y,
    keepOriginIndices: true,
  });

  if (alignedFrame) {
    alignedFrame = applySpanNullsThresholds(alignedFrame);

    // append 2 null vals at minXDelta to bar series
    if (minXDelta !== Infinity) {
      alignedFrame.fields.forEach((f, fi) => {
        let vals = f.values.toArray();

        if (fi === 0) {
          let lastVal = vals[vals.length - 1];
          vals.push(lastVal + minXDelta, lastVal + 2 * minXDelta);
        } else if (isVisibleBarField(f)) {
          vals.push(null, null);
        } else {
          vals.push(undefined, undefined);
        }
      });

      alignedFrame.length += 2;
    }

    return alignedFrame;
  }

  return null;
}

export function buildScaleKey(config: FieldConfig<GraphFieldConfig>) {
  const defaultPart = 'na';

  const scaleRange = `${config.min !== undefined ? config.min : defaultPart}-${
    config.max !== undefined ? config.max : defaultPart
  }`;

  const scaleSoftRange = `${config.custom?.axisSoftMin !== undefined ? config.custom.axisSoftMin : defaultPart}-${
    config.custom?.axisSoftMax !== undefined ? config.custom.axisSoftMax : defaultPart
  }`;

  const scalePlacement = `${
    config.custom?.axisPlacement !== undefined ? config.custom?.axisPlacement : AxisPlacement.Auto
  }`;

  const scaleUnit = config.unit ?? FIXED_UNIT;

  const scaleDistribution = config.custom?.scaleDistribution
    ? getScaleDistributionPart(config.custom.scaleDistribution)
    : ScaleDistribution.Linear;

  const scaleLabel = Boolean(config.custom?.axisLabel) ? config.custom!.axisLabel : defaultPart;

  return `${scaleUnit}/${scaleRange}/${scaleSoftRange}/${scalePlacement}/${scaleDistribution}/${scaleLabel}`;
}

function getScaleDistributionPart(config: ScaleDistributionConfig) {
  if (config.type === ScaleDistribution.Log) {
    return `${config.type}${config.log}`;
  }
  return config.type;
}
