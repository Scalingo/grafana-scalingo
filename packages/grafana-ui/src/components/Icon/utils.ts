import { IconName, IconSize } from '../../types/icon';

const alwaysMonoIcons: IconName[] = ['grafana', 'favorite', 'heart-break', 'heart', 'panel-add', 'library-panel'];

export function getIconSubDir(name: IconName, type: string): string {
  if (name?.startsWith('gf-')) {
    return 'custom';
  } else if (alwaysMonoIcons.includes(name)) {
    return 'mono';
  } else if (type === 'default') {
    return 'unicons';
  } else if (type === 'solid') {
    return 'solid';
  } else {
    return 'mono';
  }
}

/* Transform string with px to number and add 2 pxs as path in svg is 2px smaller */
export function getSvgSize(size: IconSize) {
  switch (size) {
    case 'xs':
      return 12;
    case 'sm':
      return 14;
    case 'md':
      return 16;
    case 'lg':
      return 18;
    case 'xl':
      return 24;
    case 'xxl':
      return 36;
    case 'xxxl':
      return 48;
  }
}
