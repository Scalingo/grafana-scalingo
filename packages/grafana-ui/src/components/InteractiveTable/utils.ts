import { Column as RTColumn } from 'react-table';

import { ExpanderCell } from './ExpanderCell';
import { Column } from './types';

export const EXPANDER_CELL_ID = '__expander' as const;

type InternalColumn<T extends object> = RTColumn<T> & {
  visible?: (data: T[]) => boolean;
};

// Returns the columns in a "react-table" acceptable format
export function getColumns<K extends object>(columns: Array<Column<K>>): Array<InternalColumn<K>> {
  return [
    {
      id: EXPANDER_CELL_ID,
      Cell: ExpanderCell,
      disableSortBy: true,
      width: 0,
    },
    // @ts-expect-error react-table expects each column key(id) to have data associated with it and therefore complains about
    // column.id being possibly undefined and not keyof T (where T is the data object)
    // We do not want to be that strict as we simply pass undefined to cells that do not have data associated with them.
    ...columns.map((column) => ({
      id: column.id,
      accessor: column.id,
      Header: column.header || (() => null),
      sortType: column.sortType || 'alphanumeric',
      disableSortBy: !Boolean(column.sortType),
      width: column.disableGrow ? 0 : undefined,
      visible: column.visible,
      ...(column.cell && { Cell: column.cell }),
    })),
  ];
}
