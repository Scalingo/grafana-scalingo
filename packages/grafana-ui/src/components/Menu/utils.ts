/**
 * Returns where the subMenu should be positioned (left or right)
 *
 * @param element HTMLElement for the subMenu wrapper
 */
export const getPosition = (element: HTMLElement | null) => {
  if (!element) {
    return 'left';
  }

  const wrapperPos = element.parentElement!.getBoundingClientRect();
  const pos = element.getBoundingClientRect();

  if (pos.width === 0) {
    return 'left';
  }

  if (wrapperPos.right + pos.width + 10 > window.innerWidth) {
    return 'right';
  } else {
    return 'left';
  }
};
