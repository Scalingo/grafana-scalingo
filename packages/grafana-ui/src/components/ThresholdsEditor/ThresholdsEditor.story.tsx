import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { ThresholdsEditor } from './ThresholdsEditor';
import { getTheme } from '../../themes';
import { ThresholdsMode, ThresholdsConfig } from '@grafana/data';

const ThresholdsEditorStories = storiesOf('UI/ThresholdsEditor', module);
const thresholds: ThresholdsConfig = {
  mode: ThresholdsMode.Absolute,
  steps: [
    { value: -Infinity, color: 'green' },
    { value: 50, color: 'red' },
  ],
};

ThresholdsEditorStories.add('default', () => {
  return (
    <ThresholdsEditor theme={getTheme()} thresholds={{} as ThresholdsConfig} onChange={action('Thresholds changed')} />
  );
});

ThresholdsEditorStories.add('with thresholds', () => {
  return <ThresholdsEditor theme={getTheme()} thresholds={thresholds} onChange={action('Thresholds changed')} />;
});
