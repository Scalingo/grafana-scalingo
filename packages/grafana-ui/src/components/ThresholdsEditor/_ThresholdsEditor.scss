.thresholds {
  margin-bottom: 20px;
}

.thresholds-row {
  display: flex;
  flex-direction: row;
  height: 62px;
}

.thresholds-row:first-child > .thresholds-row-color-indicator {
  border-top-left-radius: $border-radius;
  border-top-right-radius: $border-radius;
  overflow: hidden;
}

.thresholds-row:last-child > .thresholds-row-color-indicator {
  border-bottom-left-radius: $border-radius;
  border-bottom-right-radius: $border-radius;
  overflow: hidden;
}

.thresholds-row-add-button {
  @include buttonBackground($btn-success-bg, $btn-success-bg-hl, #fff);

  align-self: center;
  margin-right: 5px;
  height: 24px;
  width: 24px;
  border-radius: 50%;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
  cursor: pointer;

  &:hover {
    color: $white;
  }
}

.thresholds-row-color-indicator {
  width: 10px;
  flex-shrink: 0;
}

.thresholds-row-input {
  margin-top: 44px;
  margin-left: 2px;
}

.thresholds-row-input-inner {
  display: flex;
  justify-content: center;
  flex-direction: row;
}

.thresholds-row-input-inner > *:last-child {
  border-top-right-radius: $border-radius;
  border-bottom-right-radius: $border-radius;
}

.thresholds-row-input-inner-arrow {
  align-self: center;
  width: 0;
  height: 0;
  border-top: 6px solid transparent;
  border-bottom: 6px solid transparent;
  border-right: 6px solid $input-label-border-color;
}

.thresholds-row-input-inner-value > input {
  height: $input-height;
  padding: $input-padding;
  width: 150px;
  border-top: 1px solid $input-label-border-color;
  border-bottom: 1px solid $input-label-border-color;
}

.thresholds-row-input-inner-color {
  width: 42px;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: $input-bg;
  border: 1px solid $input-label-border-color;
}

.thresholds-row-input-inner-color-colorpicker {
  border-radius: 10px;
  display: flex;
  align-items: center;
  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.25);
}

.thresholds-row-input-inner-remove {
  display: flex;
  align-items: center;
  justify-content: center;
  height: $input-height;
  padding: $input-padding;
  width: 42px;
  background-color: $input-label-bg;
  border: 1px solid $input-label-border-color;
  cursor: pointer;
}
