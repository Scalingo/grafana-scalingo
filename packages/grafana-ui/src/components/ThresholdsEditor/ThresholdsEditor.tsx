import React, { PureComponent, ChangeEvent } from 'react';
import { Threshold, sortThresholds, ThresholdsConfig, ThresholdsMode, SelectableValue } from '@grafana/data';
import { colors } from '../../utils';
import { getColorFromHexRgbOrName } from '@grafana/data';
import { ThemeContext } from '../../themes/ThemeContext';
import { Input } from '../Input/Input';
import { ColorPicker } from '../ColorPicker/ColorPicker';
import { css } from 'emotion';
import { PanelOptionsGroup } from '../PanelOptionsGroup/PanelOptionsGroup';

export interface Props {
  thresholds?: ThresholdsConfig;
  onChange: (thresholds: ThresholdsConfig) => void;
}

interface State {
  steps: ThresholdWithKey[];
}

interface ThresholdWithKey extends Threshold {
  key: number;
}

let counter = 100;

export class ThresholdsEditor extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    const steps = toThresholdsWithKey(props.thresholds);
    steps[0].value = -Infinity;

    this.state = { steps };
  }

  onAddThresholdAfter = (threshold: ThresholdWithKey) => {
    const { steps } = this.state;

    const maxValue = 100;
    const minValue = 0;

    let prev: ThresholdWithKey | undefined = undefined;
    let next: ThresholdWithKey | undefined = undefined;
    for (const t of steps) {
      if (prev && prev.key === threshold.key) {
        next = t;
        break;
      }
      prev = t;
    }

    const prevValue = prev && isFinite(prev.value) ? prev.value : minValue;
    const nextValue = next && isFinite(next.value) ? next.value : maxValue;

    const color = colors.filter(c => !steps.some(t => t.color === c))[1];
    const add = {
      value: prevValue + (nextValue - prevValue) / 2.0,
      color: color,
      key: counter++,
    };
    const newThresholds = [...steps, add];
    sortThresholds(newThresholds);

    this.setState(
      {
        steps: newThresholds,
      },
      () => this.onChange()
    );
  };

  onRemoveThreshold = (threshold: ThresholdWithKey) => {
    const { steps } = this.state;
    if (!steps.length) {
      return;
    }
    // Don't remove index 0
    if (threshold.key === steps[0].key) {
      return;
    }
    this.setState(
      {
        steps: steps.filter(t => t.key !== threshold.key),
      },
      () => this.onChange()
    );
  };

  onChangeThresholdValue = (event: ChangeEvent<HTMLInputElement>, threshold: ThresholdWithKey) => {
    const cleanValue = event.target.value.replace(/,/g, '.');
    const parsedValue = parseFloat(cleanValue);
    const value = isNaN(parsedValue) ? '' : parsedValue;

    const steps = this.state.steps.map(t => {
      if (t.key === threshold.key) {
        t = { ...t, value: value as number };
      }
      return t;
    });
    if (steps.length) {
      steps[0].value = -Infinity;
    }
    this.setState({ steps });
  };

  onChangeThresholdColor = (threshold: ThresholdWithKey, color: string) => {
    const { steps } = this.state;

    const newThresholds = steps.map(t => {
      if (t.key === threshold.key) {
        t = { ...t, color: color };
      }

      return t;
    });

    this.setState(
      {
        steps: newThresholds,
      },
      () => this.onChange()
    );
  };

  onBlur = () => {
    const steps = [...this.state.steps];
    sortThresholds(steps);
    this.setState(
      {
        steps,
      },
      () => this.onChange()
    );
  };

  onChange = () => {
    this.props.onChange(thresholdsWithoutKey(this.props.thresholds, this.state.steps));
  };

  onModeChanged = (item: SelectableValue<ThresholdsMode>) => {
    if (item.value) {
      this.props.onChange({
        ...getThresholdOrDefault(this.props.thresholds),
        mode: item.value,
      });
    }
  };

  renderInput = (threshold: ThresholdWithKey) => {
    const config = getThresholdOrDefault(this.props.thresholds);
    const isPercent = config.mode === ThresholdsMode.Percentage;

    return (
      <div className="thresholds-row-input-inner">
        <span className="thresholds-row-input-inner-arrow" />
        <div className="thresholds-row-input-inner-color">
          {threshold.color && (
            <div className="thresholds-row-input-inner-color-colorpicker">
              <ColorPicker
                color={threshold.color}
                onChange={color => this.onChangeThresholdColor(threshold, color)}
                enableNamedColors={true}
              />
            </div>
          )}
        </div>
        {!isFinite(threshold.value) ? (
          <div className="thresholds-row-input-inner-value">
            <Input type="text" value="Base" readOnly />
          </div>
        ) : (
          <>
            <div className="thresholds-row-input-inner-value">
              <Input
                type="number"
                step="0.0001"
                onChange={(event: ChangeEvent<HTMLInputElement>) => this.onChangeThresholdValue(event, threshold)}
                value={threshold.value}
                onBlur={this.onBlur}
              />
            </div>
            {isPercent && (
              <div className={css(`margin-left:-20px; margin-top:5px;`)}>
                <i className="fa fa-percent" />
              </div>
            )}
            <div className="thresholds-row-input-inner-remove" onClick={() => this.onRemoveThreshold(threshold)}>
              <i className="fa fa-times" />
            </div>
          </>
        )}
      </div>
    );
  };

  render() {
    const { steps } = this.state;

    return (
      <PanelOptionsGroup title="Thresholds">
        <ThemeContext.Consumer>
          {theme => (
            <>
              <div className="thresholds">
                {steps
                  .slice(0)
                  .reverse()
                  .map(threshold => {
                    return (
                      <div className="thresholds-row" key={`${threshold.key}`}>
                        <div className="thresholds-row-add-button" onClick={() => this.onAddThresholdAfter(threshold)}>
                          <i className="fa fa-plus" />
                        </div>
                        <div
                          className="thresholds-row-color-indicator"
                          style={{ backgroundColor: getColorFromHexRgbOrName(threshold.color, theme.type) }}
                        />
                        <div className="thresholds-row-input">{this.renderInput(threshold)}</div>
                      </div>
                    );
                  })}
              </div>
            </>
          )}
        </ThemeContext.Consumer>
      </PanelOptionsGroup>
    );
  }
}

export function thresholdsWithoutKey(
  thresholds: ThresholdsConfig | undefined,
  steps: ThresholdWithKey[]
): ThresholdsConfig {
  thresholds = getThresholdOrDefault(thresholds);

  const mode = thresholds.mode ?? ThresholdsMode.Absolute;

  return {
    mode,
    steps: steps.map(t => {
      const { key, ...rest } = t;
      return rest; // everything except key
    }),
  };
}

function getThresholdOrDefault(thresholds?: ThresholdsConfig): ThresholdsConfig {
  return thresholds ?? { steps: [], mode: ThresholdsMode.Absolute };
}

function toThresholdsWithKey(thresholds?: ThresholdsConfig): ThresholdWithKey[] {
  thresholds = getThresholdOrDefault(thresholds);

  let steps: Threshold[] = thresholds.steps || [];

  if (thresholds.steps && thresholds.steps.length === 0) {
    steps = [{ value: -Infinity, color: 'green' }];
  }

  return steps.map(t => {
    return {
      color: t.color,
      value: t.value === null ? -Infinity : t.value,
      key: counter++,
    };
  });
}
