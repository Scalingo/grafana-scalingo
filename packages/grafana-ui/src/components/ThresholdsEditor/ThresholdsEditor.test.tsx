import React, { ChangeEvent } from 'react';
import { mount } from 'enzyme';
import { GrafanaThemeType, ThresholdsMode } from '@grafana/data';
import { ThresholdsEditor, Props, thresholdsWithoutKey } from './ThresholdsEditor';
import { colors } from '../../utils/colors';
import { mockThemeContext } from '../../themes/ThemeContext';

const setup = (propOverrides?: Partial<Props>) => {
  const props: Props = {
    onChange: jest.fn(),
    thresholds: { mode: ThresholdsMode.Absolute, steps: [] },
  };

  Object.assign(props, propOverrides);

  const wrapper = mount(<ThresholdsEditor {...props} />);
  const instance = wrapper.instance() as ThresholdsEditor;

  return {
    instance,
    wrapper,
  };
};

function getCurrentThresholds(editor: ThresholdsEditor) {
  return thresholdsWithoutKey(editor.props.thresholds, editor.state.steps);
}

describe('Render', () => {
  let restoreThemeContext: any;
  beforeAll(() => {
    restoreThemeContext = mockThemeContext({ type: GrafanaThemeType.Dark });
  });

  afterAll(() => {
    restoreThemeContext();
  });

  it('should render with base threshold', () => {
    const { wrapper } = setup();
    expect(wrapper.find('.thresholds')).toMatchSnapshot();
  });
});

describe('Initialization', () => {
  it('should add a base threshold if missing', () => {
    const { instance } = setup();
    expect(getCurrentThresholds(instance).steps).toEqual([{ value: -Infinity, color: 'green' }]);
  });
});

describe('Add threshold', () => {
  it('should add threshold', () => {
    const { instance } = setup();

    instance.onAddThresholdAfter(instance.state.steps[0]);

    expect(getCurrentThresholds(instance).steps).toEqual([
      { value: -Infinity, color: 'green' }, // 0
      { value: 50, color: colors[1] }, // 1
    ]);
  });

  it('should add another threshold above a first', () => {
    const { instance } = setup({
      thresholds: {
        mode: ThresholdsMode.Absolute,
        steps: [
          { value: -Infinity, color: colors[0] }, // 0
          { value: 50, color: colors[2] }, // 1
        ],
      },
    });

    instance.onAddThresholdAfter(instance.state.steps[1]);

    expect(getCurrentThresholds(instance).steps).toEqual([
      { value: -Infinity, color: colors[0] }, // 0
      { value: 50, color: colors[2] }, // 1
      { value: 75, color: colors[3] }, // 2
    ]);
  });

  it('should add another threshold between first and second index', () => {
    const { instance } = setup({
      thresholds: {
        mode: ThresholdsMode.Absolute,
        steps: [
          { value: -Infinity, color: colors[0] },
          { value: 50, color: colors[2] },
          { value: 75, color: colors[3] },
        ],
      },
    });

    instance.onAddThresholdAfter(instance.state.steps[1]);

    expect(getCurrentThresholds(instance).steps).toEqual([
      { value: -Infinity, color: colors[0] },
      { value: 50, color: colors[2] },
      { value: 62.5, color: colors[4] },
      { value: 75, color: colors[3] },
    ]);
  });
});

describe('Remove threshold', () => {
  it('should not remove threshold at index 0', () => {
    const thresholds = {
      mode: ThresholdsMode.Absolute,
      steps: [
        { value: -Infinity, color: '#7EB26D' },
        { value: 50, color: '#EAB839' },
        { value: 75, color: '#6ED0E0' },
      ],
    };
    const { instance } = setup({ thresholds });

    instance.onRemoveThreshold(instance.state.steps[0]);

    expect(getCurrentThresholds(instance)).toEqual(thresholds);
  });

  it('should remove threshold', () => {
    const thresholds = {
      mode: ThresholdsMode.Absolute,
      steps: [
        { value: -Infinity, color: '#7EB26D' },
        { value: 50, color: '#EAB839' },
        { value: 75, color: '#6ED0E0' },
      ],
    };
    const { instance } = setup({ thresholds });

    instance.onRemoveThreshold(instance.state.steps[1]);

    expect(getCurrentThresholds(instance).steps).toEqual([
      { value: -Infinity, color: '#7EB26D' },
      { value: 75, color: '#6ED0E0' },
    ]);
  });
});

describe('change threshold value', () => {
  it('should not change threshold at index 0', () => {
    const thresholds = {
      mode: ThresholdsMode.Absolute,
      steps: [
        { value: -Infinity, color: '#7EB26D' },
        { value: 50, color: '#EAB839' },
        { value: 75, color: '#6ED0E0' },
      ],
    };
    const { instance } = setup({ thresholds });

    const mockEvent = ({ target: { value: '12' } } as any) as ChangeEvent<HTMLInputElement>;

    instance.onChangeThresholdValue(mockEvent, instance.state.steps[0]);

    expect(getCurrentThresholds(instance)).toEqual(thresholds);
  });

  it('should update value', () => {
    const { instance } = setup();
    const thresholds = {
      mode: ThresholdsMode.Absolute,
      steps: [
        { value: -Infinity, color: '#7EB26D', key: 1 },
        { value: 50, color: '#EAB839', key: 2 },
        { value: 75, color: '#6ED0E0', key: 3 },
      ],
    };

    instance.state = {
      steps: thresholds.steps,
    };

    const mockEvent = ({ target: { value: '78' } } as any) as ChangeEvent<HTMLInputElement>;

    instance.onChangeThresholdValue(mockEvent, thresholds.steps[1]);

    expect(getCurrentThresholds(instance).steps).toEqual([
      { value: -Infinity, color: '#7EB26D' },
      { value: 78, color: '#EAB839' },
      { value: 75, color: '#6ED0E0' },
    ]);
  });
});

describe('on blur threshold value', () => {
  it('should resort rows and update indexes', () => {
    const { instance } = setup();
    const thresholds = {
      mode: ThresholdsMode.Absolute,
      steps: [
        { value: -Infinity, color: '#7EB26D', key: 1 },
        { value: 78, color: '#EAB839', key: 2 },
        { value: 75, color: '#6ED0E0', key: 3 },
      ],
    };

    instance.setState({
      steps: thresholds.steps,
    });

    instance.onBlur();

    expect(getCurrentThresholds(instance).steps).toEqual([
      { value: -Infinity, color: '#7EB26D' },
      { value: 75, color: '#6ED0E0' },
      { value: 78, color: '#EAB839' },
    ]);
  });
});
