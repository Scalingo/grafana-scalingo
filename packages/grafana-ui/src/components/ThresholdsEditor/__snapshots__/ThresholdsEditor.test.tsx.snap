// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should render with base threshold 1`] = `
<div
  className="thresholds"
>
  <div
    className="thresholds-row"
    key="100"
  >
    <div
      className="thresholds-row-add-button"
      onClick={[Function]}
    >
      <i
        className="fa fa-plus"
      />
    </div>
    <div
      className="thresholds-row-color-indicator"
      style={
        Object {
          "backgroundColor": "#73BF69",
        }
      }
    />
    <div
      className="thresholds-row-input"
    >
      <div
        className="thresholds-row-input-inner"
      >
        <span
          className="thresholds-row-input-inner-arrow"
        />
        <div
          className="thresholds-row-input-inner-color"
        >
          <div
            className="thresholds-row-input-inner-color-colorpicker"
          >
            <WithTheme(ColorPicker)
              color="green"
              enableNamedColors={true}
              onChange={[Function]}
            >
              <ColorPicker
                color="green"
                enableNamedColors={true}
                onChange={[Function]}
                theme={
                  Object {
                    "type": "dark",
                  }
                }
              >
                <PopoverController
                  content={
                    <ColorPickerPopover
                      color="green"
                      enableNamedColors={true}
                      onChange={[Function]}
                      theme={
                        Object {
                          "type": "dark",
                        }
                      }
                    />
                  }
                  hideAfter={300}
                >
                  <ForwardRef(ColorPickerTrigger)
                    color="#73BF69"
                    onClick={[Function]}
                    onMouseLeave={[Function]}
                  >
                    <div
                      onClick={[Function]}
                      onMouseLeave={[Function]}
                      style={
                        Object {
                          "background": "inherit",
                          "border": "none",
                          "borderRadius": 10,
                          "color": "inherit",
                          "cursor": "pointer",
                          "overflow": "hidden",
                          "padding": 0,
                        }
                      }
                    >
                      <div
                        style={
                          Object {
                            "backgroundImage": "url(data:image/png,base64,iVBORw0KGgoAAAANSUhEUgAAAAwAAAAMCAIAAADZF8uwAAAAGUlEQVQYV2M4gwH+YwCGIasIUwhT25BVBADtzYNYrHvv4gAAAABJRU5ErkJggg==)",
                            "border": "none",
                            "float": "left",
                            "height": 15,
                            "margin": 0,
                            "position": "relative",
                            "width": 15,
                            "zIndex": 0,
                          }
                        }
                      >
                        <div
                          style={
                            Object {
                              "backgroundColor": "#73BF69",
                              "bottom": 0,
                              "display": "block",
                              "left": 0,
                              "position": "absolute",
                              "right": 0,
                              "top": 0,
                            }
                          }
                        />
                      </div>
                    </div>
                  </ForwardRef(ColorPickerTrigger)>
                </PopoverController>
              </ColorPicker>
            </WithTheme(ColorPicker)>
          </div>
        </div>
        <div
          className="thresholds-row-input-inner-value"
        >
          <Input
            className=""
            readOnly={true}
            type="text"
            value="Base"
          >
            <div
              style={
                Object {
                  "flexGrow": 1,
                }
              }
            >
              <input
                className="gf-form-input"
                readOnly={true}
                type="text"
                value="Base"
              />
            </div>
          </Input>
        </div>
      </div>
    </div>
  </div>
</div>
`;
